heat/core/linalg/svdtools.py

```diff
@@ -0,0 +1,532 @@
+"""
+distributed hierarchical SVD - first draft
+"""
+import numpy as np
+import collections
+import torch
+from typing import Type, Callable, Dict, Any, TypeVar, Union, Tuple
+
+from ..communication import MPICommunication
+from ..dndarray import DNDarray
+from .. import factories
+from .. import types
+from ..linalg import matmul, vector_norm
+from ..indexing import where
+from ..random import randn
+
+from ..manipulations import vstack, hstack, diag, balance
+
+from .. import statistics
+from math import log, ceil, floor, sqrt
+
+
+__all__ = ["hsvd_rank", "hsvd_reltol", "hsvd"]
+
+
+#########################################################################################
+# user-friendly versions of hSVD
+#########################################################################################
+
+
+def hsvd_rank(
+    A: DNDarray,
+    maxrank: int,
+    maxmergedim: Union[int, None] = None,
+    safetyshift: int = 5,
+    full: bool = False,
+    silent: bool = True,
+) -> Union[
+    Tuple[DNDarray, DNDarray, DNDarray, float], Tuple[DNDarray, DNDarray, DNDarray], DNDarray
+]:
+    """
+    Hierarchical SVD (hSVD) with prescribed truncation rank maxrank.
+    If A = U diag(sigma) V^T is the true SVD of A, this routine computes an approximation for U[:,:maxrank] (and sigma[:maxrank], V[:,:maxrank])
+
+    The accuracy of this approximation depends on the structure of A ("low-rank" is best) and appropriate choice of parameters.
+
+    Parameters
+    ----------
+    A : DNDarray
+        2D-array (float32/64) of which the hSVD has to be computed.
+    maxrank : int
+        truncation rank.
+    full : bool, optional
+        full=True implies that also Sigma and V are computed and returned. The default is False.
+
+    'Expert' parameters
+    ----------
+
+    maxmergedim : Union[int, None], optional
+        maximal size of the concatenation matrices during the merging procedure. The default is None and results in an appropriate choice depending on the size of the local slices of A and maxrank.
+    safetyshift : int, optional
+        Increases the actual truncation rank within the computations by a safety shift. The default is 5.
```

Comment:
I'm trying to understand whether "safety shift" is an established term in truncated matrix decomposition, in which case it will be obvious, apologies. 

Would it make sense to express the safety shift in % of truncation rank?

Reply:
I have no issue with the method. I worry that the initial user effort to map sklearn TruncatedSVD functionalities to hSVD is very significant, and will result in low adoption. @mrfh92 if you agree let's  solve this offline.